cdc/log.cc

```diff
@@ -654,154 +715,111 @@ class transformer final {
+            throw std::runtime_error(format("cdc: empty mutation? {}", base_mut));
+        }
 
-        auto& p = m.partition();
-        if (p.partition_tombstone()) {
-            // Partition deletion
-            auto log_ck = set_pk_columns(m.key(), ts, tuuid, 0, res);
-            set_operation(log_ck, ts, operation::partition_delete, res);
-        } else if (!p.row_tombstones().empty()) {
-            // range deletion
+        auto stream_id = get_stream(changes, base_mut.token());
+        mutation log_mut(_log_schema, stream_id.to_partition_key(*_log_schema));
+
+        for (auto& [time, btch]: changes) {
+            auto tuuid = timeuuid_type->decompose(generate_timeuuid(time));
             int batch_no = 0;
-            for (auto& rt : p.row_tombstones()) {
-                auto set_bound = [&] (const clustering_key& log_ck, const clustering_key_prefix& ckp) {
-                    auto exploded = ckp.explode(*_schema);
-                    size_t pos = 0;
-                    for (const auto& column : _schema->clustering_key_columns()) {
-                        if (pos >= exploded.size()) {
-                            break;
-                        }
-                        auto cdef = _log_schema->get_column_definition(to_bytes("_" + column.name()));
-                        auto value = atomic_cell::make_live(*column.type,
-                                                            ts,
-                                                            bytes_view(exploded[pos]),
-                                                            _cdc_ttl_opt);
-                        res.set_cell(log_ck, *cdef, std::move(value));
-                        ++pos;
-                    }
-                };
-                {
-                    auto log_ck = set_pk_columns(m.key(), ts, tuuid, batch_no, res);
-                    set_bound(log_ck, rt.start);
-                    // TODO: separate inclusive/exclusive range
-                    set_operation(log_ck, ts, operation::range_delete_start, res);
-                    ++batch_no;
-                }
-                {
-                    auto log_ck = set_pk_columns(m.key(), ts, tuuid, batch_no, res);
-                    set_bound(log_ck, rt.end);
-                    // TODO: separate inclusive/exclusive range
-                    set_operation(log_ck, ts, operation::range_delete_end, res);
-                    ++batch_no;
+
+            for (auto& change: btch.static_updates) {
+                // TODO: find a way to get a preimage for the static row, even if there are no rows
+                // corresponding to clustering keys affected by the mutation (in particular,
+                // even if the mutation affects no clustering rows at all)
+                // from CQL static columns can be queried specifying only the partition key in the WHERE clause
+                // --- it will return values for static columns even if there are no clustering rows
```

Comment:
Any read to a partition using the internal APIs, using any slice will also return the static row.

Reply:
I checked experimentally: if you specify a clustering key in the bounds, and there is no clustered row with this clustering key, the static row won't be returned.

At least not from
```
    future<coordinator_query_result> storage_proxy::query(schema_ptr,
        lw_shared_ptr<query::read_command> cmd,
        dht::partition_range_vector&& partition_ranges,
        db::consistency_level cl,
        coordinator_query_options optional_params);
```